lib/src/fortio.cpp

```diff
@@ -0,0 +1,337 @@
+                std::uint16_t x;
+                std::memcpy( &x, ptr, sizeof( x ) );
+                x = ntohs( x );
+                std::memcpy( ptr, &x, sizeof( x ) );
+                ptr += 2;
+            }
+        } else {
+            for( int i = 0; i < head / o.elemsize; ++i ) {
+                std::uint16_t x;
+                std::memcpy( &x, ptr, sizeof( x ) );
+                x = ntohl( x );
+                std::memcpy( ptr, &x, sizeof( x ) );
+                ptr += 4;
+            }
+        }
+    }
+
+    if( o.force_notail ) {
+        if( recordsize ) *recordsize = elems;
+        guard.fp = nullptr;
+        return ECL_OK;
+    }
+
+    fguard tailguard( fp );
+    std::int32_t tail;
+    read = std::fread( &tail, sizeof( tail ), 1, fp );
+
+    if( read == 1 and o.bigendian )
+        tail = (std::int32_t)ntohl( (std::uint32_t)tail );
+
+    /*
+     * success; record has a tail, and it matched our head. All is good -
+     * return and don't rewind. This is the most common case
+     */
+    if( read == 1 and head == tail ) {
+        if( recordsize ) *recordsize = elems;
+        tailguard.fp = guard.fp = nullptr;
+        return ECL_OK;
+    }
+
+    /*
+     * read was sucessful, but the tail didn't match. However, missing tails
+     * are ok, so rewind the tail, and preserve position at end-of-record
+     */
+    if( read == 1 and head != tail and o.allow_notail ) {
+        if( recordsize ) *recordsize = elems;
+        guard.fp = nullptr;
+        return ECL_OK;
+    }
+
+    /*
+     * last record, and ok to not have tail. don't rewind as this is a success
+     */
+    if( read != 1 and feof( fp ) and o.allow_notail ) {
+        if( recordsize ) *recordsize = elems;
+        guard.fp = tailguard.fp = nullptr;
+        return ECL_OK;
+    }
+
+    /*
+     * read was succesful, but the tail didn't match, so flag the record as
+     * invalid and rewind
+     */
+    if( read == 1 and head != tail )
+        return ECL_INVALID_RECORD;
+
+    /* read failed, but not at end-of-file */
+    if( read != 1 and ferror( fp ) )
+        return ECL_ERR_READ;
+
+    /*
+     * The read failed at end-of-file, and we expected a tail. Flag the record
+     * as invalid and roll back the file pointer
+     */
+    if( read != 1 and feof( fp ) and not o.allow_notail )
+        return ECL_INVALID_RECORD;
+
+    return ECL_ERR_UNKNOWN;
+} catch( std::runtime_error& ) {
+    return ECL_ERR_SEEK;
+}
+
+int eclfio_put( std::FILE* fp,
+                const char* opts,
+                int nmemb,
+                const void* record ) try {
+
+    const auto o = parse_opts( opts );
+
+    if(   std::numeric_limits< std::int32_t >::max()
+        < std::int64_t(nmemb) * std::int64_t(o.elemsize) ) {
+        /*
+         * a huge record (>2G) is being requested, but the 4-byte headers don't
+         * allow that. consider this an invalid argument and bail
+         */
+        return ECL_EINVAL;
+    }
+
+    fguard guard( fp );
```

Comment:
Could we scope the guard variable:
```C++
{
      fguard guard(fp);
      ....
      ....
}
return ECL_OK;
```
then we can let the destructor play out normally - without the explcit `guard = nullptr` assignment?

Reply:
Well, you need to signal the guard that we're exiting on success, otherwise it'll attempt to roll back.

What *could* be done is a trivial method that reads return value:

```
int exit( int errc ) { if( errc == ECL_OK ) this->fp = nullptr; return errc; }

[...]

if( head == tail ) {
    return guard.exit( ECL_OK );
}
```

But that's only slightly more ergonomic.

Reply:
Fair enough - I realized after I wrote that I misunderstood the mechanism slightly.